### ISSUE TITLE
RTF-file produced by `esttab` cannot display UTF-8 characters correctly in MS-word when using stata14 or newer version

### ISSUE THREAD

**Issue by dreistein543** [benjann/estout#34]
This problem has been existing for a long time (since the update of encoding mode in stata14), and I used to avoid this problem by not using non-ASCII words in my command before.

`estout.pkg` is an amazing work and has been put out before UTF-8 been used in Stata. At that time, for non-ASCII users (eg. Chinese users), can output their results through `esttab` with non-ASCII characters because the ANSI encoding mode is flexible.

Things changed that Stata use UTF-8 encoding mode. And thank you Ben, `estout.pkg` has been added UTF-8 support, as you declared. but the RTF-file still cannot specify these non-ASCII characters correctly. Similar problems is available from: [Stata Forum](https://www.statalist.org/forums/forum/general-stata-discussion/general/1325981-encoding-issue-in-rtf-with-esttab-esttout-under-stata-14)

I, as a Chinese user, have tried the `unicode convertfile` methods, it doesn't work for Chinese characters.
besides, I tried to use `ustrto()` to convert Chinese characters to old gb2312(or gb18030) encoding mode, which used in Stata13. and add these converted characters into `esttab` syntax. still not working.

it fails me a lot and push me to seek help from you.

P.S. non-ASCII characters like Chinese characters do display well in **Stata windows**, but cannot display correctly in **RTF-file opened by MS-word**.

A few simple code with Chinese character may help you to check if this problem fixed.
```stata
sysuse auto, clear
eststo m1: reg price length
```
```stata
esttab m1, title(回归结果)
* correct in Stata window

回归结果
----------------------------
                      (1)   
                    price   
----------------------------
length              57.20***
                   (4.06)   

_cons             -4584.9   
                  (-1.72)   
----------------------------
N                      74   
----------------------------
t statistics in parentheses
* p<0.05, ** p<0.01, *** p<0.001
```

```stata
esttab m1 using myTable.rtf, replace title(回归结果) 

* uncorrect in RTF-file as the picture follows
```
<img width="360" alt="截屏2022-03-23 01 29 27" src="https://user-images.githubusercontent.com/30745407/159541481-bb299d0e-09da-4fb7-a756-5c7abbb6ee3d.png">

**Comment by NilsEnevoldsen** [comment]
First define:

```stata
program define enc_rtf, rclass
version 16.1
syntax, str(string)
local rtf_str ""
local hex_str = ustrtohex("`str'")
local hex_chars = subinstr("`hex_str'","\u"," ",.)
foreach hex_char of local hex_chars {
	quietly inten 16 `hex_char'
	local rtf_str = `"`rtf_str'\u`r(ten)'?"'
}
return local rtf_str `"`rtf_str'"'
end
```

Then:

```stata
sysuse auto, clear
eststo m1: reg price length
enc_rtf, str("回归结果")
esttab m1 using ~/myTable.rtf, replace title(`r(rtf_str)') 
```

Something like this could be done internally, but up to Ben if it's important enough.

**Comment by NilsEnevoldsen** [comment]
An explanation of the above:

1. Start with `回归结果`. 
2. `ustrtohex()` transforms it into `\u56de\u5f52\u7ed3\u679c`.
3. Eliminate `\u`s to get ` 56de 5f52 7ed3 679c`.
4. Transform this from hexidecimal to decimal: `22238 24402 32467 26524`
5. Add back the `\u`s, and add a fallback character (`?`) after each codepoint: `\u22238?\u24402?\u32467?\u26524?`

This is the final Unicode escape format used by RTF.

I didn't see an elegant way to do this with base Stata commands, but such a way might exist.

**Comment by benjann** [comment]
Or here's a Mata function doing the same:

```
mata:
    string scalar enc_rtf(string scalar s)
    {
        return(invtokens("\u" :+
            strofreal(frombase(16,tokens(subinstr(ustrtohex(s), "\u", " "))))
            :+ "?", ""))
    }
end
```

Could then use it as follows:

```
sysuse auto, clear
eststo m1: reg price length
mata: st_local("title", enc_rtf("回归结果"))
esttab m1 using ~/myTable.rtf, replace title(`title')
```

The function is not very efficient (although more efficient than the ado version provided by Nils). What about `ustrto()`? Does this maybe provide a translator to RTF-compatible unicode?

**Comment by NilsEnevoldsen** [comment]
I don't think any of the available encodings are "Unicode, decimal escaped", which is I think what we would need.

```stata
. unicode encoding list *unicode*

   #                      Encoding                                        Alias
-------------------------------------------------------------------------------
   1                         UTF-8                            unicode-1-1-utf-8
                                                              unicode-2-0-utf-8
   2                        UTF-16                                      unicode
                                                                      csUnicode
   3                      UTF-16BE                           UnicodeBigUnmarked
   4                      UTF-16LE                        UnicodeLittleUnmarked
   5            UTF-16BE,version=1                                   UnicodeBig
   6            UTF-16LE,version=1                                UnicodeLittle
   7                         UTF-7                            unicode-1-1-utf-7
                                                              unicode-2-0-utf-7
   8            ibm-5471_P100-2006                        big5-hkscs:unicode3.0
-------------------------------------------------------------------------------
```

**Comment by dreistein543** [comment]
Thank you Nils and Ben, your `enc_rtf` ado and meta-function do solve this problem when specifying a single string.
`esttab` has some special areas need a string list to customize my RTF-file like:
```stata
esttab [models] using ~/myTable.rtf, stat(r2 N, label("R方" "样本量"))
```
It works if I specify these strings separately, but that is not efficient and elegant. you know mostly, we use macros to do this incase we need to change the stats-info later:

```stata
local title "回归结果" // regression result
local note "括号中为 t 统计量" // t statistics in parentheses
local statList "r2 N"
local labList `"R方" "样本量"' // `"R-square"  "Observations"'
esttab [models] using ~/myTable.rtf, ///
      stat(`statList',label(`labList')) title(`title') note(`note')

* as for the local title & note, I can do as follows simply:
foreach str in title note {
    enc_rtf, str("``str''")
    local `str' `r(rtf_str)'
// or meta function
    meta: st_local("`str'", enc_rtf("``str''"))
}
* then I can use the locals `title' and `note' directly.
* but when it comes to string list `labList', 
* enc_rtf ado/meta function may not batch process them in a simple way
```

**Comment by benjann** [comment]
Define:

```Stata
mata:
    string scalar enc_rtf(string scalar s)
    {
        real scalar      i
        real rowvector   c
        string rowvector S
        
        c = frombase(16, tokens(subinstr(ustrtohex(s), "\u", " ")))
        i = length(c)
        S = J(1,i,"")
        for (;i;i--) {
            if (c[i]<=127) S[i] = char(c[i])
            else           S[i] = "\u" + strofreal(c[i]) + "?"
        }
        return(invtokens(S, ""))
    }
end
```

and then use as

```Stata
sysuse auto, clear
eststo m1: reg price length
mata: st_local("title", enc_rtf("回归结果"))
mata: st_local("note", enc_rtf("括号中为 t 统计量"))
local statList "r2 N"
mata: st_local("labList", enc_rtf(`""R方" "样本量""'))
esttab m1 using ~/myTable.rtf, replace ///
      stat(`statList', label(`labList')) title(`title') note(`note')
```

Function `enc_rtf()` leaves ASCII characters (code <= 127) as is and only translates non-ASCII characters (code > 127).

Detail: In your example it should be 
```Stata
local labList `""R方" "样本量""'
```
not
```Stata
local labList `"R方" "样本量"'
```
Note the difference in use of double quotes. However, since the two elements do not contain spaces, it could also simply be
```Stata
local labList "R方 样本量"
```

ben

**Comment by benjann** [comment]
I now posted an update that automatically escapes non-ASCII characters in RTF mode in Stata 14 or newer. Old behavior can be restored by option `nortfencode`.

That is, you can now type
```
esttab m1 using ~/myTable.rtf, ///
      stat(r2 N, label("R方" "样本量")) title(回归结果) note(括号中为 t 统计量)
```
and the table should display fine.

**Comment by dreistein543** [comment]
such an amazing news! Thanks for your efforts, Ben!

**Comment by benjann** [comment]
The new feature is currently only available in the version of `estout` hosted at GitHub; it may take a while until the change will make it into the version available from the SSC Archive. So, for now, use
```
 net install estout, replace from(https://raw.githubusercontent.com/benjann/estout/master/)
```
to obtain the update (rather than `ssc install estout`).

**Comment by dreistein543** [comment]
this problem may not close so fast, a new bug occur when I try to specify no-ASCII labels in `cells()`:

```stata
sysuse auto, clear
eststo: tab: estpost tab foreign
esttab tab using ~/myTable.rtf, replace cells((b(label(频数)) /// 
                                          pct(fmt(2) label("频率")) ///
                                       cumpct(fmt(2) label("加和")))) ///
                                       noobs nomtitle nonumber
```

it can run with no warnings or error.
but when I open the RTF-file, MS-word warns me that “Word cannot read this file. file maybe damaged.”

**Comment by benjann** [comment]
Not good. It seems the line containing the labels gets truncated. I'll try to fix this.

**Comment by benjann** [comment]
The reason is that Mata's `ustrtohex()` has a 200 character limit; I somehow missed that. I will program a workaround.

**Comment by dreistein543** [comment]
just in case.

when specify `mtitle()` `mgroup()` and `indicate(,label())`, same problem occurs.

`estout.pkg` is such a huge work. it is me only use non-ASCII characters in
 
- [x] `title()`,

- [x]  `note()`, 

- [ ] `cells(stat(label([label])))` 

- [x] `stat(,label([label]))`, 

- [x] `indicate([Name]=)` eg. specify `indicate(年份=*.year)` to diplay i.year ✓

- [ ] `indicate(, label([mark]))` eg. specify `indicate(Year = *.year, label("是" "否"))` doesn't work

- [ ]  `mtitle([model List])` 

- [ ] `mgroup()` 

- [ ] `other variables' label`

- [ ] `other variables' name`

- [x] `all variables' value label`

- [x] `depVar's name` 

- [x] `depVar's variable/value label` (depVar's value label may use in `mlogit` but **canno**t use with `unstack`)

An interesting phenomenon is if depVar is a _binominal_ variable, `esttab` can output `mlogit` results with `unstack` correct (P.S. it is meanless to use mlogit model on a binominal depVar of course), but if depVar contains **three or more values**, the RTF-file won't open by MS-word

I marked those can display correct with the most current version 

I'm not sure if there are other function may use non-ASCII strings.

**Comment by benjann** [comment]
I now posted a fix; lines longer than 200 characters should now be processed correctly so that the resulting RTF will no longer be corrupt. (Note that apparently RTF can only handle Unicode up to character 65535; `estout` will set larger codes to Unicode replacement character 65533).

**Comment by benjann** [comment]
Also note that the Unicode translation is applied at the end, after the complete RTF code has been written. It should hence not matter in which options you use non-ASCII characters.

**Comment by NilsEnevoldsen** [comment]
Does it handle characters between 32767 and 65535? The [RTF spec](https://www.biblioscape.com/rtf15_spec.htm#Heading9) says "Unicode values greater than 32767 must be expressed as negative numbers".

**Comment by benjann** [comment]
The RTF readers I checks had no problem with that (MS Word and TextEdit on Mac). An alternative would be the use \u with negative codes computed as `#-65536` for numbers larger than 32767. Maybe this would be a safer approach (?).

**Comment by NilsEnevoldsen** [comment]
Here's the current [1.9.1 spec](https://interoperability.blob.core.windows.net/files/Archive_References/%5bMSFT-RTF%5d.pdf). It does seem to specify that `#-65536` would be safer. But I personally have no idea in practice.

**Comment by benjann** [comment]
Thanks. To be on the safe side, I now changed the translator so that `#-65536` is used for codes >32767.

**Comment by NilsEnevoldsen** [comment]
@benjann This is fixed in 3.29, which is available on SSC. It would be appropriate to close this issue.
